File: backend/crud/worlds.py
"""
CRUD operations for World entities.

Handles world creation, retrieval, updates, and deletion.
"""

import json
import logging
from datetime import datetime, timezone
from typing import List, Optional

import schemas
from domain.value_objects.enums import WorldPhase
from infrastructure.database import models
from infrastructure.database.connection import serialized_write
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload

logger = logging.getLogger("WorldCRUD")

# Gameplay agents that should be added to location rooms
# Only Action_Manager and Narrator are in the tape - sub-agents (Item_Designer,
# Character_Designer, Location_Designer) are invoked via tools, not added to rooms
GAMEPLAY_AGENT_NAMES = [
    "Action_Manager",
    "Narrator",
]


async def get_gameplay_agents(db: AsyncSession) -> List[models.Agent]:
    """Get all gameplay agents by name."""
    result = await db.execute(select(models.Agent).where(models.Agent.name.in_(GAMEPLAY_AGENT_NAMES)))
    return list(result.scalars().all())


async def add_gameplay_agents_to_room(db: AsyncSession, room_id: int) -> int:
    """
    Add all gameplay agents to a location room.

    Args:
        db: Database session
        room_id: ID of the room to add agents to

    Returns:
        Number of agents added
    """
    from crud.room_agents import add_agent_to_room

    agents = await get_gameplay_agents(db)
    added_count = 0

    for agent in agents:
        result = await add_agent_to_room(db, room_id, agent.id)
        if result:
            added_count += 1
            logger.debug(f"Added gameplay agent '{agent.name}' to room {room_id}")

    if added_count > 0:
        logger.info(f"Added {added_count} gameplay agents to room {room_id}")

    return added_count


async def create_world(
    db: AsyncSession,
    world: schemas.WorldCreate,
    owner_id: str,
) -> models.World:
    """Create a new world with an onboarding room and empty player state."""
    from crud.rooms import create_room

    # Check if onboarding room already exists (from failed previous attempt)
    onboarding_room_name = f"Onboarding: {world.name}"
    result = await db.execute(
        select(models.Room).where(
            models.Room.owner_id == owner_id,  # Scoped to world owner
            models.Room.name == onboarding_room_name,
        )
    )
    onboarding_room = result.scalar_one_or_none()

    # Create onboarding room if it doesn't exist
    if onboarding_room is None:
        onboarding_room = await create_room(
            db,
            schemas.RoomCreate(name=onboarding_room_name),
            owner_id=owner_id,  # Use world owner_id to scope rooms per world owner
        )
    else:
        logger.info(f"Reusing existing onboarding room: {onboarding_room_name}")

    db_world = models.World(
        name=world.name,
        owner_id=owner_id,
        user_name=world.user_name,
        language=world.language,
        phase=WorldPhase.ONBOARDING,
        onboarding_room_id=onboarding_room.id,
    )
    db.add(db_world)
    await db.flush()

    # Link onboarding room to world (for CASCADE delete)
    onboarding_room.world_id = db_world.id

    # Create empty player state
    player_state = models.PlayerState(
        world_id=db_world.id,
        turn_count=0,
        stats="{}",
        inventory="[]",
        effects="[]",
        action_history="[]",
    )
    db.add(player_state)

    async with serialized_write():
        await db.commit()

    await db.refresh(db_world, attribute_names=["locations", "player_state", "onboarding_room"])
    return db_world


async def get_world(db: AsyncSession, world_id: int) -> Optional[models.World]:
    """Get a world by ID with relationships."""
    result = await db.execute(
        select(models.World)
        .options(
            selectinload(models.World.locations),
            selectinload(models.World.player_state).selectinload(models.PlayerState.current_location),
            selectinload(models.World.onboarding_room),
        )
        .where(models.World.id == world_id)
    )
    return result.scalar_one_or_none()


async def get_world_by_name(
    db: AsyncSession,
    name: str,
    owner_id: str,
) -> Optional[models.World]:
    """Get a world by name and owner."""
    result = await db.execute(
        select(models.World).where(models.World.name == name).where(models.World.owner_id == owner_id)
    )
    return result.scalar_one_or_none()


async def get_worlds_by_owner(
    db: AsyncSession,
    owner_id: str,
) -> List[models.World]:
    """Get all worlds owned by a user, sorted by last activity."""
    result = await db.execute(
        select(models.World)
        .where(models.World.owner_id == owner_id)
        .order_by(models.World.last_played_at.desc().nullsfirst())
    )
    return list(result.scalars().all())


async def update_world(
    db: AsyncSession,
    world_id: int,
    update: schemas.WorldUpdate,
) -> Optional[models.World]:
    """Update world configuration."""
    result = await db.execute(select(models.World).where(models.World.id == world_id))
    world = result.scalar_one_or_none()

    if not world:
        return None

    if update.phase is not None:
        world.phase = update.phase
    if update.genre is not None:
        world.genre = update.genre
    if update.theme is not None:
        world.theme = update.theme
    if update.user_name is not None:
        world.user_name = update.user_name
    if update.stat_definitions is not None:
        world.stat_definitions = json.dumps(update.stat_definitions)

    world.updated_at = datetime.now(timezone.utc)

    async with serialized_write():
        await db.commit()

    await db.refresh(world)
    return world


async def update_world_last_played(
    db: AsyncSession,
    world_id: int,
) -> None:
    """Update the last_played_at timestamp."""
    result = await db.execute(select(models.World).where(models.World.id == world_id))
    world = result.scalar_one_or_none()

    if world:
        world.last_played_at = datetime.now(timezone.utc)
        async with serialized_write():
            await db.commit()


async def delete_world(db: AsyncSession, world_id: int) -> bool:
    """Delete a world and all associated data.

    In PostgreSQL, rooms are deleted via Room.world_id FK CASCADE.
    For SQLite (tests), we manually delete rooms to ensure cleanup.
    Also deletes all agents associated with the world.
    """
    result = await db.execute(select(models.World).where(models.World.id == world_id))
    world = result.scalar_one_or_none()

    if not world:
        return False

    world_name = world.name

    # Get all rooms associated with this world (for manual deletion in SQLite)
    rooms_result = await db.execute(select(models.Room).where(models.Room.world_id == world_id))
    rooms_to_delete = rooms_result.scalars().all()

    # Get all agents associated with this world (for deletion)
    agents_result = await db.execute(select(models.Agent).where(models.Agent.world_name == world_name))
    agents_to_delete = agents_result.scalars().all()

    # Break circular dependency: World.onboarding_room_id <-> Room.world_id
    world.onboarding_room_id = None
    await db.flush()

    # Delete all world-specific agents
    for agent in agents_to_delete:
        await db.delete(agent)
    if agents_to_delete:
        logger.info(f"Deleted {len(agents_to_delete)} agents for world '{world_name}'")

    # Delete the world - cascades to:
    # - locations (via World.locations cascade)
    # - player_state (via World.player_state cascade)
    await db.delete(world)
    await db.flush()

    # Manually delete rooms (needed for SQLite which doesn't enforce FK CASCADE)
    for room in rooms_to_delete:
        await db.delete(room)

    async with serialized_write():
        await db.commit()

    return True


async def import_world_from_filesystem(
    db: AsyncSession,
    fs_config,  # WorldConfig from domain.world_models
    owner_id: str,
) -> models.World:
    """
    Import a world from filesystem into the database.

    This creates:
    - A World record in database with filesystem metadata
    - A placeholder onboarding room (for structural consistency)
    - Empty player state

    The world's phase, genre, theme etc. come from filesystem config.
    """
    from services.location_service import LocationService
    from services.player_service import PlayerService

    from crud.rooms import create_room

    # Create a room for the world (needed for DB structure)
    # For active worlds, this is the "home" room
    room_name = f"World: {fs_config.name}"
    room = await create_room(
        db,
        schemas.RoomCreate(name=room_name),
        owner_id=owner_id,
    )

    # Determine phase from filesystem
    phase = fs_config.phase
    if isinstance(phase, str):
        phase = WorldPhase(phase)

    # Create the world record
    db_world = models.World(
        name=fs_config.name,
        owner_id=owner_id,
        user_name=fs_config.user_name,
        language=fs_config.language,
        phase=phase,
        genre=fs_config.genre,
        theme=fs_config.theme,
        onboarding_room_id=room.id,
        created_at=fs_config.created_at or datetime.now(timezone.utc),
        updated_at=fs_config.updated_at or datetime.now(timezone.utc),
    )
    db.add(db_world)
    await db.flush()

    # Link room to world (for CASCADE delete)
    room.world_id = db_world.id

    # Load player state from filesystem
    fs_player_state = PlayerService.load_player_state(fs_config.name)

    # Create player state from filesystem data
    import json

    player_state = models.PlayerState(
        world_id=db_world.id,
        turn_count=fs_player_state.turn_count if fs_player_state else 0,
        stats=json.dumps(fs_player_state.stats) if fs_player_state else "{}",
        inventory=json.dumps(fs_player_state.inventory) if fs_player_state else "[]",
        effects=json.dumps(fs_player_state.effects) if fs_player_state else "[]",
        action_history=json.dumps(fs_player_state.recent_actions) if fs_player_state else "[]",
    )
    db.add(player_state)

    # Store room mapping in _state.json for FS-first architecture
    if phase == WorldPhase.ACTIVE:
        LocationService.set_room_mapping(
            world_name=fs_config.name,
            room_key="main",
            db_room_id=room.id,
            agents=["Action_Manager", "Narrator"],
        )
    else:
        LocationService.set_room_mapping(
            world_name=fs_config.name,
            room_key="onboarding",
            db_room_id=room.id,
            agents=["Onboarding_Manager"],
        )

    async with serialized_write():
        await db.commit()

    await db.refresh(db_world, attribute_names=["locations", "player_state", "onboarding_room"])

    logger.info(f"Imported world '{fs_config.name}' from filesystem (phase={phase})")
    return db_world

File: backend/src/crud/worlds.ts
/** CRUD operations for World entities. */

import { and, asc, eq, inArray, isNull, sql } from 'drizzle-orm'
import type { Db } from '@/db'
import {
  agents,
  locations,
  playerStates,
  rooms,
  worlds,
  WORLD_PHASES,
  type Agent,
  type Language,
  type Location,
  type Room,
  type World,
  type WorldPhase,
} from '@/db/schema'
import { invalidateRoomCache } from './cache-invalidation'
import type { PlayerStateWithLocation } from './player-state'
import { addAgentToRoom, createRoom } from './rooms'
import { toLangKey } from '@/domain/enums'
import { PlayerStateSerializer } from '@/domain/player-state-serializer'
import { getLogger } from '@/infrastructure/logging/logger'
import type { PlayerService } from '@/services/player-service'
import type { RoomMappingService } from '@/services/room-mapping'
import type { WorldConfig } from '@/services/world-service'

const logger = getLogger('WorldCRUD')

// Only the two tape participants; the designer sub-agents are dispatched
// through the SDK instead, and room membership would only add noise.
export const GAMEPLAY_AGENT_NAMES = ['Action_Manager', 'Narrator'] as const

export interface WorldWithRelations extends World {
  locations: Location[]
  playerState: PlayerStateWithLocation | null
  onboardingRoom: Room | null
}

export interface WorldCreate {
  name: string
  /** The player's display name *inside* the world, not a user id. */
  userName?: string | null
  language?: Language
}

// One transaction, because a world without a player state cannot be rendered.
// The callback keeps using `db` rather than `tx` on purpose: `bun:sqlite` has
// one connection, so {@link createRoom} joins this transaction as it stands.
export function createWorld(db: Db, world: WorldCreate, ownerId: string): WorldWithRelations {
  const onboardingRoomName = `Onboarding: ${world.name}`

  const worldId = db.transaction(() => {
    // Owner-scoped: names are unique per (owner, name, world_id), so a global
    // lookup could adopt another tenant's room as this world's.
    const existingRoom =
      db
        .select()
        .from(rooms)
        .where(and(eq(rooms.ownerId, ownerId), eq(rooms.name, onboardingRoomName)))
        .get() ?? null

    if (existingRoom) logger.info(`Reusing existing onboarding room: ${onboardingRoomName}`)
    const onboardingRoom = existingRoom ?? createRoom(db, { name: onboardingRoomName }, ownerId)

    // No SQL DEFAULT and no `$defaultFn`: omitting these writes NULLs into the
    // columns the world list sorts on.
    const now = new Date()
    const created = db
      .insert(worlds)
      .values({
        name: world.name,
        ownerId,
        userName: world.userName ?? null,
        language: world.language ?? 'en',
        phase: 'onboarding',
        onboardingRoomId: onboardingRoom.id,
        createdAt: now,
        updatedAt: now,
      })
      .returning()
      .get()

    // The FK cycle (`worlds.onboarding_room_id` ↔ `rooms.world_id`) needs two
    // statements, so the reverse link is set after the insert.
    db.update(rooms).set({ worldId: created.id }).where(eq(rooms.id, onboardingRoom.id)).run()

    // Literal `"{}"` / `"[]"`, not the `jsonOrNull` convention of `messages.ts`.
    db.insert(playerStates)
      .values({
        worldId: created.id,
        turnCount: 0,
        stats: '{}',
        inventory: '[]',
        effects: '[]',
        actionHistory: '[]',
      })
      .run()

    return created.id
  })

  // Non-null: the row was just inserted in a committed transaction.
  return getWorld(db, worldId)!
}

export function getWorld(db: Db, worldId: number): WorldWithRelations | null {
  const world = db.select().from(worlds).where(eq(worlds.id, worldId)).get()
  if (!world) return null

  const playerStateRow = db
    .select({ state: playerStates, currentLocation: locations })
    .from(playerStates)
    .leftJoin(locations, eq(playerStates.currentLocationId, locations.id))
    .where(eq(playerStates.worldId, worldId))
    .get()

  return {
    ...world,
    locations: db
      .select()
      .from(locations)
      .where(eq(locations.worldId, worldId))
      .orderBy(asc(locations.id))
      .all(),
    playerState: playerStateRow
      ? { ...playerStateRow.state, currentLocation: playerStateRow.currentLocation }
      : null,
    onboardingRoom:
      world.onboardingRoomId === null
        ? null
        : (db.select().from(rooms).where(eq(rooms.id, world.onboardingRoomId)).get() ?? null),
  }
}

// `ownerId` is required: names are unique per *owner* (`ux_worlds_owner_name`),
// so an ownerless lookup is a cross-tenant read. `null` means "the ownerless
// worlds" and needs the `IS NULL` branch — `eq(col, null)` matches nothing.
export function getWorldByName(db: Db, name: string, ownerId: string | null): World | null {
  const where = and(
    eq(worlds.name, name),
    ownerId === null ? isNull(worlds.ownerId) : eq(worlds.ownerId, ownerId),
  )!
  return db.select().from(worlds).where(where).get() ?? null
}

// Most recently played first. `NULLS FIRST` is explicit because SQLite sorts
// NULL lowest, burying a freshly created world under every old save.
export function getWorldsByOwner(db: Db, ownerId: string): World[] {
  return db
    .select()
    .from(worlds)
    .where(eq(worlds.ownerId, ownerId))
    .orderBy(sql`${worlds.lastPlayedAt} desc nulls first`)
    .all()
}

export interface WorldUpdate {
  phase?: WorldPhase | null
  genre?: string | null
  theme?: string | null
  userName?: string | null
  statDefinitions?: Record<string, unknown> | null
}

// `undefined`/`null` both mean "leave alone", so a field cannot be cleared back
// to NULL here. An empty `statDefinitions` is stored as `"{}"`, keeping "no stat
// system" (NULL) distinguishable from "configured as empty".
export function updateWorld(db: Db, worldId: number, update: WorldUpdate): World | null {
  const world = db.select().from(worlds).where(eq(worlds.id, worldId)).get()
  if (!world) return null

  const patch: Partial<typeof worlds.$inferInsert> = {
    // Always stamped, even when nothing else changed.
    updatedAt: new Date(),
  }
  if (update.phase !== undefined && update.phase !== null) patch.phase = update.phase
  if (update.genre !== undefined && update.genre !== null) patch.genre = update.genre
  if (update.theme !== undefined && update.theme !== null) patch.theme = update.theme
  if (update.userName !== undefined && update.userName !== null) patch.userName = update.userName
  if (update.statDefinitions !== undefined && update.statDefinitions !== null) {
    patch.statDefinitions = JSON.stringify(update.statDefinitions)
  }

  return db.update(worlds).set(patch).where(eq(worlds.id, worldId)).returning().get() ?? null
}

export function updateWorldLastPlayed(db: Db, worldId: number): void {
  db.update(worlds).set({ lastPlayedAt: new Date() }).where(eq(worlds.id, worldId)).run()
}

export function getGameplayAgents(db: Db): Agent[] {
  return db
    .select()
    .from(agents)
    .where(inArray(agents.name, [...GAMEPLAY_AGENT_NAMES]))
    .all()
}

// Idempotent. The count is *not* "how many were newly added" — it feeds a log.
export function addGameplayAgentsToRoom(db: Db, roomId: number): number {
  let count = 0
  for (const agent of getGameplayAgents(db)) {
    if (addAgentToRoom(db, roomId, agent.id)) count += 1
  }
  return count
}

// The order orphans rows rather than failing when it is wrong: break the FK
// cycle (`worlds.onboarding_room_id` ↔ `rooms.world_id`), then world-scoped
// agents while `world_name` is known, then the world, then its rooms — that
// last step is redundant under `PRAGMA foreign_keys=ON` but nothing else.
export function deleteWorld(db: Db, worldId: number): boolean {
  const world = db.select().from(worlds).where(eq(worlds.id, worldId)).get()
  if (!world) return false

  const roomIds = db
    .select({ id: rooms.id })
    .from(rooms)
    .where(eq(rooms.worldId, worldId))
    .all()
    .map((r) => r.id)

  db.transaction((tx) => {
    tx.update(worlds).set({ onboardingRoomId: null }).where(eq(worlds.id, worldId)).run()

    const deletedAgents = tx
      .delete(agents)
      .where(eq(agents.worldName, world.name))
      .returning({ id: agents.id })
      .all()
    if (deletedAgents.length > 0) {
      logger.info(`Deleted ${deletedAgents.length} agents for world '${world.name}'`)
    }

    tx.delete(worlds).where(eq(worlds.id, worldId)).run()

    if (roomIds.length > 0) tx.delete(rooms).where(inArray(rooms.id, roomIds)).run()
  })

  // SQLite reuses rowids: a later room inheriting a deleted one's id would be
  // served its cached `is_paused` / `is_finished`, which gate the scheduler.
  for (const roomId of roomIds) invalidateRoomCache(roomId)

  return true
}

// `players` reads `player.json`; `rooms` writes the mapping into `_state.json`.
export interface ImportWorldServices {
  players: PlayerService
  rooms: RoomMappingService
}

// `world.json` is user-editable; throwing beats defaulting to `onboarding`,
// which would restart a finished campaign.
function toWorldPhase(phase: string): WorldPhase {
  if ((WORLD_PHASES as readonly string[]).includes(phase)) return phase as WorldPhase
  throw new Error(`Unknown world phase: ${phase}`)
}

// The `_state.json` write sits *inside* the transaction, so a failure rolls the
// rows back rather than leaving a database claiming a room nothing maps.
export function importWorldFromFilesystem(
  db: Db,
  fsConfig: WorldConfig,
  ownerId: string,
  services: ImportWorldServices,
): WorldWithRelations {
  const phase = toWorldPhase(fsConfig.phase)

  const worldId = db.transaction(() => {
    // `db` rather than `tx`; see {@link createWorld}.
    const room = createRoom(db, { name: `World: ${fsConfig.name}` }, ownerId)

    const created = db
      .insert(worlds)
      .values({
        name: fsConfig.name,
        ownerId,
        userName: fsConfig.userName,
        language: toLangKey(fsConfig.language),
        phase,
        genre: fsConfig.genre,
        theme: fsConfig.theme,
        onboardingRoomId: room.id,
        // The config loader already defaults these to "now" when the YAML omits them.
        createdAt: fsConfig.createdAt,
        updatedAt: fsConfig.updatedAt,
      })
      .returning()
      .get()

    db.update(rooms).set({ worldId: created.id }).where(eq(rooms.id, room.id)).run()

    // A world can exist on disk with no `player.json` yet.
    const fsPlayer = services.players.loadPlayerState(fsConfig.name)
    db.insert(playerStates)
      .values({
        worldId: created.id,
        turnCount: fsPlayer?.turnCount ?? 0,
        stats: fsPlayer ? PlayerStateSerializer.serializeStats(fsPlayer.stats) : '{}',
        inventory: fsPlayer ? PlayerStateSerializer.serializeInventory(fsPlayer.inventory) : '[]',
        effects: fsPlayer ? PlayerStateSerializer.serializeEffects(fsPlayer.effects) : '[]',
        // `recent_actions` on disk, `action_history` in the column.
        actionHistory: fsPlayer
          ? PlayerStateSerializer.serializeActionHistory(fsPlayer.recentActions)
          : '[]',
      })
      .run()

    // The key differs by phase because the casts do: gameplay tape vs interviewer.
    if (phase === 'active') {
      services.rooms.setRoomMapping(fsConfig.name, 'main', room.id, [...GAMEPLAY_AGENT_NAMES])
    } else {
      services.rooms.setRoomMapping(fsConfig.name, 'onboarding', room.id, ['Onboarding_Manager'])
    }

    return created.id
  })

  logger.info(`Imported world '${fsConfig.name}' from filesystem (phase=${phase})`)
  return getWorld(db, worldId)!
}
